### PR TITLE
[MIG] base_location_nuts: Migration to 10.0

### PR DESCRIPTION
**Pull request by chienandalu**
cc @Tecnativa

### CODE REVIEW

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: I'm pretty sure this kind of api is not available in v10. Please check that and use `@api.onchange` instead if possible.

File: base_location_nuts/views/res_partner_view.xml
Comment (yajo) [inline base_location_nuts/views/res_partner_view.xml]: `width=1%` does something?

File: base_location_nuts/views/res_partner_view.xml
Comment (yajo) [inline base_location_nuts/views/res_partner_view.xml]: Same on width

File: base_location_nuts/views/res_partner_view.xml
Comment (yajo) [inline base_location_nuts/views/res_partner_view.xml]: Same on width

File: base_location_nuts/views/res_partner_view.xml
Comment (yajo) [inline base_location_nuts/views/res_partner_view.xml]: Same on width.

File: base_location_nuts/views/res_partner_view.xml
Comment (yajo) [inline base_location_nuts/views/res_partner_view.xml]: newline

File: base_location_nuts/wizard/nuts_import.py
Comment (yajo) [inline base_location_nuts/wizard/nuts_import.py]: Import it as `UserError`; [`Warning` is deprecated.](https://github.com/odoo/odoo/blob/10.0/odoo/exceptions.py#L36-L37)

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: Don't you need `oldname` in these cols? Or any other migrations?

**Review by yajo — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
Please rebase to get a green branch

**Comment by pedrobaeza** [comment]
Waiting fixes before reviewing

File: base_location_nuts/README.rst
Comment (pedrobaeza) [inline base_location_nuts/README.rst]: https

File: base_location_nuts/__manifest__.py
Comment (pedrobaeza) [inline base_location_nuts/__manifest__.py]: Remove Antiun

File: base_location_nuts/__manifest__.py
Comment (pedrobaeza) [inline base_location_nuts/__manifest__.py]: Comma at the end

File: base_location_nuts/__manifest__.py
Comment (pedrobaeza) [inline base_location_nuts/__manifest__.py]: https

File: base_location_nuts/models/res_partner.py
Comment (pedrobaeza) [inline base_location_nuts/models/res_partner.py]: Please use `getattr(self, "nuts%_id")`

File: base_location_nuts/models/res_partner.py
Comment (pedrobaeza) [inline base_location_nuts/models/res_partner.py]: Don't use `.id`. You can compare directly record against record

File: base_location_nuts/models/res_partner.py
Comment (pedrobaeza) [inline base_location_nuts/models/res_partner.py]: Use `getattr`

File: base_location_nuts/models/res_partner.py
Comment (pedrobaeza) [inline base_location_nuts/models/res_partner.py]: This method is not used anymore. Please find the equivalent one or declare a new unique onchange (`def onchange_state_id_base_location_nuts(self):`.

File: base_location_nuts/models/res_partner_nuts.py
Comment (pedrobaeza) [inline base_location_nuts/models/res_partner_nuts.py]: There's no need of oldname in one2many fields, as they don't have a real column in the DB.

File: base_location_nuts/views/res_partner_view.xml
Comment (pedrobaeza) [inline base_location_nuts/views/res_partner_view.xml:10]: Why do you remove the `attrs` attribute?

File: base_location_nuts/wizard/nuts_import.py
Comment (pedrobaeza) [inline base_location_nuts/wizard/nuts_import.py]: Old alias. Now `UserError` is the valid one.

File: base_location_nuts/wizard/nuts_import.py
Comment (pedrobaeza) [inline base_location_nuts/wizard/nuts_import.py]: Put `...` at the end to indicate that is an operation that requires time.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: base_location_nuts/views/res_partner_view.xml
Comment (chienandalu) [reply base_location_nuts/views/res_partner_view.xml:10]: I didn't. That change comes from a previous commit (https://github.com/Tecnativa/partner-contact/commit/eade50cd3ab61b354346e8bb6e227508d59008c3)

**Comment by pedrobaeza** [comment]
Please increase test coverage for the onchanges and so on.

**Comment by chienandalu** [comment]
I was doing some test and was getting 404 from the NUTS Eurostat server. There's some issue with Ramon not accepting https request. It does it right on http though.

**Comment by pedrobaeza** [comment]
Well, change it then.

**Comment by chienandalu** [comment]
@Yajo @pedrobaeza It's all green now

**Review by pedrobaeza — APPROVED** [review]
Remember to squash your commits before merging

**Comment by chienandalu** [comment]
Commits squashed

**Comment by chienandalu** [comment]
Yes, @pedrobaeza, my fault for copy-pasting and not passing flake locally prior to commit :grimacing: 
Can you review again, @Yajo ?

**Comment by pedrobaeza** [comment]
Please put pre-commit linter in your git for avoiding these "surprises": https://github.com/OCA/maintainer-quality-tools/tree/master/git

**Comment by chienandalu** [comment]
@pedrobaeza Thanks!

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: Do not use getattr, use `self["nuts%d_id" % level]` instead.

Se slides 40 and 41: https://www.odoo.com/es_ES/slides/slide/10-rules-for-safer-code-398

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: 😕  Why not just `if level > 1`?

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: Why not `parent_id = field.parent_id`?

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: no getattr for accessing fields.

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: All these `api.multi` can be dropped, now in v10 it's the default.

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: `level = int(field[4])`

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: Not needed

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py:82]: You can do `for field in fields[1:]` and remove the `fields.remove` part

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: 🗑

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: `self.country_id = self.state_id.country_id`

File: base_location_nuts/models/res_partner.py
Comment (yajo) [inline base_location_nuts/models/res_partner.py]: Remove lines 117-119, they are redundant. If you edit `self` under an onchange in the new api, you don't need to return a dict with values.

File: base_location_nuts/models/res_partner_nuts.py
Comment (yajo) [inline base_location_nuts/models/res_partner_nuts.py]: Add `oldname='children'`

File: base_location_nuts/wizard/nuts_import.py
Comment (yajo) [inline base_location_nuts/wizard/nuts_import.py]: Wrong indentation here?

File: base_location_nuts/wizard/nuts_import_view.xml
Comment (yajo) [inline base_location_nuts/wizard/nuts_import_view.xml:38]: It's sad to have to add the `sales_team` dependency just for putting the menu somewhere. Do you think there's another place where we can put it and skip this big dependency just for that? 🤔 Or maybe we really need this for other reason?

**Review by yajo — CHANGES_REQUESTED** [review]
Humm complex module, some suggestions to simplify it in some parts 😊 

BTW, there are some security good practices that it's good to follow, see below.

Thanks for the good job!

**Comment by chienandalu** [comment]
@Yajo Thanks for your comments :) I've tried to implement them all.

File: base_location_nuts/wizard/nuts_import_view.xml
Comment (chienandalu) [reply base_location_nuts/wizard/nuts_import_view.xml:38]: This family of modules fall into that menu historically. It's no really needed at all for anything else. There aren't much options on base views anyway... Ideas are welcomed :smile:

Comment (pedrobaeza) [reply base_location_nuts/wizard/nuts_import_view.xml:38]: Well, on base_location there is another menu defined: https://github.com/OCA/partner-contact/blob/10.0/base_location/views/better_zip_view.xml#L61

Comment (chienandalu) [reply base_location_nuts/wizard/nuts_import_view.xml:38]: Right, @pedrobaeza , but this module doesn't depend on that one :confused: It only depended on base before...

Comment (pedrobaeza) [reply base_location_nuts/wizard/nuts_import_view.xml:38]: Uhm, OK, let's go this way then.

File: base_location_nuts/models/res_partner.py
Comment (yajo) [reply base_location_nuts/models/res_partner.py]: No, he is using the right thing. Attributes can get anything from the object, while keys can only get fields, so for security this is the recommended system to access fields with variable names, and even more when the variable comes from outside the method.

Please check slide 40 of https://www.odoo.com/es_ES/slides/slide/10-rules-for-safer-code-398

File: base_location_nuts/wizard/nuts_import.py
Comment (yajo) [inline base_location_nuts/wizard/nuts_import.py]: The indentation problem I was referring to is that these are 8 spaces instead of the usual 4. (I mean from the line 195 to 196).

**Review by yajo — APPROVED** [review]

**Comment by yajo** [comment]
@chienandalu this looks almost perfect now. Could you please fix that little comment and squash all the latest migration-related commits to proceed with merge?

File: base_location_nuts/models/res_partner.py
Comment (pedrobaeza) [reply base_location_nuts/models/res_partner.py]: Well, this system can't be abused like the scenario of the slides. My fear is that this is not a documented way of accessing a field. The official way is with dot notation (and thus, `getattr`), but I see in the slides that is using this method, so go on...

**Comment by chienandalu** [comment]
@Yajo Fixed and commits squashed

File: base_location_nuts/models/res_partner.py
Comment (yajo) [reply base_location_nuts/models/res_partner.py]: Well, that presentation was given by @odony, so I guess that makes it somehow officially documented. Maybe a PR to docs would be good.

Comment (odony) [reply base_location_nuts/models/res_partner.py]: :+1: let's make it more official: odoo/odoo@464fe4a185316cf9af9ae15385bd55662534109b

Comment (pedrobaeza) [reply base_location_nuts/models/res_partner.py]: Thank your very much for the update on the documentation, Olivier!

Comment (yajo) [reply base_location_nuts/models/res_partner.py]: Thanks, great little addition! 👏 👏

**Pull request merged**
